change_operations/insert_operation.py

```diff
@@ -0,0 +1,384 @@
+    # Now process the normalized dependencies
+    for (before, after), dep in normalized_deps.items():
+        if after == cur_activity:
+            if before in visited:
+                raise RecursionError(f"Temporal contradiction: cycle detected between '{before}' and '{cur_activity}'")
+            visited.add(before)
+            visited = visited | dfs(temporal_deps, before, visited)
+    return visited
+
+def has_temporal_contradiction(
+    temporal_deps: Dict[Tuple[str, str], TemporalDependency],
+    existential_deps: Dict[Tuple[str, str], ExistentialDependency],
+    activities: List[str],
+    activity: str,
+    variants: List[List[str]],
+):
+    """
+    Checks if there is a contradiction in the temporal dependencies.
+
+    Args:
+        temporal_deps: The temporal deps for which the check should be performed.
+        existential_deps: The existential dependencies belonging to the temporal_deps.
+        activities: List of all activities.
+        activity: List of activity to be inserted
+        variants: Variants of the matrix for which the check should be performed.
+
+    Returns:
+        True, if it has a contradiction
+        False, if there is no contradiction
+    """
+    after = set()
+    before = set()
+    for source, target in temporal_deps:
+        dep = temporal_deps.get((source, target))
+        if dep.type != TemporalType.DIRECT:
+            continue
+        
+        # Determine actual order based on direction
+        if dep.direction == Direction.FORWARD:
+            # source comes before target
+            actual_before, actual_after = source, target
+        elif dep.direction == Direction.BACKWARD:
+            # target comes before source (source comes after target)
+            actual_before, actual_after = target, source
+        else:  # Direction.BOTH - skip for contradiction checking
+            continue
+            
+        if actual_after == activity:
+            before.add(actual_before)
+        if actual_before == activity:
+            after.add(actual_after)
+
+    for variant in variants:
+        variant_activities = set(variant)
+        variant_activities.add(activity)
+        if not satisfies_existential_constraints(variant_activities, activities, existential_deps):
+            continue
+
+        n = 0
+        b_pos = -1
+        for b in before:
+            if b in variant:
+                b_pos = variant.index(b)
+                n += 1
+            if n > 1:
+                return True
+        n = 0
+        a_pos = -1
+        for a in after:
+            if a in variant:
+                a_pos = variant.index(a)
+                n += 1
+            if n > 1:
+                return True
+        if (a_pos != -1) and (b_pos != -1):
+            if (a_pos != (b_pos + 1)):
+                return True
+
+    try:
+        for a in activities:
+            dfs(temporal_deps, a, set())
+    except RecursionError as e:
+        return True
+
+    return False
```

Comment:
this should also be it `utils/` and make the dfs function private (in python it just means u start the function name with an underscore so it becomes `_dfs`